pkg/epp/datalayer/metrics/client.go

```diff
@@ -0,0 +1,135 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/expfmt"
+	"sigs.k8s.io/gateway-api-inference-extension/pkg/epp/datalayer"
+)
+
+// Client is an interface for retrieving the metrics from an endpoint URL.
+type Client interface {
+	Get(ctx context.Context, target *url.URL, ep datalayer.Addressable) (PrometheusMetricMap, error)
+}
+
+// ClientFactory returns a Client suitable for an endpoint.
+// Implementations may return a new Client each time, or use a cached
+// copy for optimized retrieval.
+type ClientFactory interface {
+	GetClientForEndpoint(ep datalayer.Addressable) Client
+}
+
+// GetDefaultClientFactory returns a default implementation of the
+// ClientFactory, which caches and reuses client across calls.
+func GetDefaultClientFactory() ClientFactory {
+	return defaultClientFactory
+}
+
+// -- package implementations --
+var (
+	cleanupTick          = 30 * time.Second
+	maxIdleTime          = time.Minute
+	defaultClientFactory = newClientFactory()
+)
+
+type client struct {
+	cl       *http.Client
+	lastUsed time.Time
+}
+
+type clientmap struct {
+	cleanupOnce sync.Once
+	clients     sync.Map // map[string]*client (uses client IP as key)
+}
+
+func newClientFactory() *clientmap {
+	clm := &clientmap{}
+	clm.startCleanupGoroutine()
+	return clm
+}
+
+func (clm *clientmap) startCleanupGoroutine() {
+	clm.cleanupOnce.Do(func() {
+		go func() {
+			for {
+				time.Sleep(cleanupTick)
+				now := time.Now()
+
+				clm.clients.Range(func(key, value any) bool {
+					entry := value.(*client)
+					if now.Sub(entry.lastUsed) > maxIdleTime {
+						clm.clients.Delete(key)
+					}
+					return true
+				})
+			}
+		}()
+	})
+}
+
+func (clm *clientmap) GetClientForEndpoint(ep datalayer.Addressable) Client {
+	cl := newClient()
+	id := ep.GetIPAddress()
+
+	if value, loaded := clm.clients.LoadOrStore(id, cl); !loaded {
```

Comment:
the creation of newClient for each endpoint every 50 ms seems a bit problematic.
GC might take time and memory may increase more and more with unused client objects (every cycle a new client object is created and then not used cause you read from the sync.Map).
it might be better to first check if it exists in the map before creation (even though it may add few code liens)